### PR TITLE
Go based installer 

### PR DESCRIPTION
**Pull request by Naatan**
https://www.pivotaltracker.com/story/show/179332688

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -284,7 +345,7 @@ jobs:
       run: |
         export SHELL=''
         set +e
-        go test -timeout 10m github.com/ActiveState/cli/test/integration -test.v -test.run "^TestLanguageMsiActivePerl|TestDeployIntegrationTestSuite|TestInitIntegrationTestSuite|TestPushIntegrationTestSuite|TestUpdateIntegrationTestSuite|PrepareIntegrationTestSuite|InstallScriptsIntegrationTestSuite|ShowIntegrationTestSuite|PerformanceIntegrationTestSuite$"
+        go test -timeout 10m `go list ./... | grep integration` -test.v -test.run "^TestLanguageMsiActivePerl|TestDeployIntegrationTestSuite|TestInitIntegrationTestSuite|TestPushIntegrationTestSuite|TestUpdateIntegrationTestSuite|PrepareIntegrationTestSuite|InstallScriptsIntegrationTestSuite|ShowIntegrationTestSuite|PerformanceIntegrationTestSuite$"
```
Comment (mdrohmann) [inline .github/workflows/build.yml:348]: Just a suggestion:
```suggestion
        TEST_SUITE_TAGS=critical:msi:deploy:init:push:update:prepare:installer:install-scripts:show:performance go test -timeout 10m `go list ./... | grep integration` -test.v
```

If we move to that scheme, we can become a little bit more selective in the future by adding only selected tests from a files to the critical path.

**Review by mdrohmann — CHANGES_REQUESTED** [review]
Mostly nit comments and suggestions, but what really needs to be addressed is: ensure that we fetch with `source=install` when we invoke the installer from an install script.

File: installers/install.sh [hunk 2]
```diff
@@ -55,194 +20,71 @@ if [ "$SESSION_TOKEN" != "$SESSION_TOKEN_VERIFY" ]; then
   SESSION_TOKEN_VALUE=$SESSION_TOKEN
 fi
 
+while getopts "b:" arg; do
+  case $arg in
+    b) CHANNEL=$OPTARG;;
+  esac
+done
+
 if [ -z "${TERM}" ] || [ "${TERM}" = "dumb" ]; then
```
[the remark this answers is not in the document]
Comment (Naatan) [reply installers/install.sh:29]: I'll file a followup. `tput` isn't something new, it's been in these scripts for ages far as I can tell.

[the remark this answers is not in the document]
Comment (Naatan) [reply installers/install.sh:29]: https://www.pivotaltracker.com/story/show/179690536

File: installers/install.ps1 [hunk 3]
```diff
@@ -111,316 +69,89 @@ function download([string] $url, [string] $out) {
...
-        & $executable $arguments
-    }
-    else {
-        # Keep --activate and --activate-default flags for backwards compatibility
-        activateIfRequested
-    }
-}
-exit $code
+$env:ACTIVESTATE_SESSION_TOKEN = $script:SESSION_TOKEN_VALUE
+& $exePath $args
```
[the remark this answers is not in the document]
Comment (Naatan) [reply installers/install.ps1]: Resolved with https://github.com/ActiveState/cli/pull/1521/commits/5a983b2a1667a1bc25d1f18b3c73c83b1f21caae

File: cmd/state-installer/installer.go [hunk 4]
```diff
@@ -1,196 +1,147 @@
...
-			return errs.Wrap(err, "Failed to retrieve absolute installPath")
-		}
-	} else {
-		installPath, err = installation.InstallPath()
-		if err != nil {
-			return errs.Wrap(err, "Failed to retrieve default installPath")
+	// Store update tag
+	if i.updateTag != "" {
+		if err := i.cfg.Set(updater.CfgUpdateTag, i.updateTag); err != nil {
+			return errs.Wrap(err, "Failed to set update tag")
```
[the remark this answers is not in the document]
Comment (Naatan) [reply cmd/state-installer/installer.go:52]: No, sanitize() is just sanitizing the input data, it's not storing any config.

File: .github/workflows/build.yml (hunk 1 above)
Comment (Naatan) [reply .github/workflows/build.yml:348]: I like this approach, but I don't want to complicate this PR any further with potential fallout from an unrelated change.

https://www.pivotaltracker.com/story/show/179693154

**Review by mdrohmann — APPROVED** [review]

**Comment by Naatan** [comment]
@mdrohmann you should be able to review with this range:

https://github.com/ActiveState/cli/pull/1521/files/6dacec2253880ffd3fa814d332953abee84f4c02..4ea433bf90e4283b16c72e538eb0489d7d268a66

**Review by mdrohmann — CHANGES_REQUESTED** [review]
Nice work on the word-wrap refactor.  That makes a lot more sense this way.

File: internal/appinfo/appinfo.go [hunk 5]
```diff
@@ -18,6 +19,10 @@ type AppInfo struct {
 
 func execDir(baseDir ...string) string {
 	if len(baseDir) > 0 {
+		binDir := filepath.Join(baseDir[0], "bin")
+		if fileutils.DirExists(binDir) {
+			return binDir
+		}
```
[the remark this answers is not in the document]
Comment (Naatan) [reply internal/appinfo/appinfo.go:25]: The whole bin dir is pretty awkward atm. There are a number of places where this is sort of hard coded. I've filed a followup to clean up this particular aspect, which isn't exactly new in this PR, just ran into it more: https://activestatef.atlassian.net/browse/ST-1115

**Review by mdrohmann — APPROVED** [review]

**Pull request merged**